gdk/x11/gdkglcontext-x11-eglx.c

```diff
@@ -810,13 +843,28 @@ gdk_x11_display_make_gl_context_current (GdkDisplay   *display,
 
   if (context_x11->is_attached)
     {
-      /* If the WM is compositing there is no particular need to delay
-       * the swap when drawing on the offscreen, rendering to the screen
-       * happens later anyway, and its up to the compositor to sync that
-       * to the vblank. */
       GdkScreen *screen = gdk_window_get_screen (window);
+      gboolean is_composited = gdk_screen_is_composited (screen);
+
+#ifdef HAVE_XCOMPOSITE
+      GdkToplevelX11 *toplevel = _gdk_x11_window_get_toplevel (window);
+
+      /* If the WM is compositing and the window is redirected there is no
+       * particular need to delay the swap when drawing on the offscreen,
+       * rendering to the screen happens later anyway, and its up to the
+       * compositor to sync that to the vblank.
+       */
+      if (is_composited)
+        {
+          if (toplevel && gdk_x11_screen_supports_net_wm_hint (screen,
+              gdk_atom_intern_static_string ("_GTK_WINDOW_UNREDIRECTED")))
+            is_composited = !toplevel->unredirected;
+          else if (gdk_window_get_state (window) == GDK_WINDOW_STATE_FULLSCREEN)
+            is_composited = window_is_composited (display, window->impl_window);
+        }
+#endif /* HAVE_XCOMPOSITE */
```

Comment:
Instead of a big `#ifdef HAVE_XCOMPOSITE … #endif` block, you should probably always define `window_is_composited()` and just move the `#ifdef HAVE_XCOMPOSITE` block inside the function itself. This way the function existence is not conditional, but its body is. If we don't have the XComposite API available, then we should always return `FALSE`, e.g.:

```c
static gboolean
window_is_composited (GdkDisplay *display, GdkWindow *window)
{
#ifdef HAVE_XCOMPOSITE
  // Do the XCompositeNameWindowPixmap() call
#else
  return FALSE;
#endif
}